Add Image.Extent() interface

nirs opened the pull request:
Extent() return the next extent in the image, starting at the specified
offset, and limited by the specified length. An extent is a run of
clusters of with the same status.
    
Using Extent() we can iterate over the image and skip extents that reads
as zeros, instead of reading a zero buffer and detecting that the buffer
is full of zeros.
    
Benchmarking show 2 orders of magnitude speeded compared with zero
detection.
  
```console
% go test -bench Benchmark0p/qcow2$/read
Benchmark0p/qcow2/read-12      516     2503752 ns/op      107213.28 MB/s    1050537 B/op    39 allocs/op
    
% go test -bench NextUnallocated
BenchmarkNextUnallocated-12    116    10236186 ns/op    10489666.91 MB/s          0 B/op     0 allocs/op
```

Ubuntu 24.04 image in qcow2 compressed format:

```console
% ./go-qcow2reader-example convert /var/tmp/images/test.zlib.qcow2 /tmp/tmp.img
6.00 GiB / 6.00 GiB [-------------------------------------] 100.00% 2.86 GiB p/s 
```

Ubuntu 24.04 image in qcow2 format:

```console
% ./go-qcow2reader-example convert /var/tmp/images/test.qcow2 /tmp/tmp.img
6.00 GiB / 6.00 GiB [------------------------------------] 100.00% 15.73 GiB p/s 
```

16 TiB empty qcow2 image:

```console
% ./go-qcow2reader-example convert /var/tmp/images/test.0p.qcow2 /tmp/tmp.img
16.00 TiB / 16.00 TiB [---------------------------------] 100.00% 152.46 TiB p/s 
```

Fixes #32

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -959,6 +959,103 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+	}
+
+	// Do not allow requests after the end of the image. This is likely a bug in
+	// the caller code, so better to fail fast and keep the code simpler.
```
This does not work for the case we test:
```
base image (2g) <- top image (4g)
```

Read from child from offset 2g to 4g should return unallocated cluster - we get the same result when reading after the end of the image.

nirs commented on image/stub/stub.go:
```diff
@@ -25,6 +25,11 @@ type Stub struct {
 
+// Next returns the next extent.
+func (img *Stub) Next(start, length int64) (image.Extent, error) {
+	return image.Extent{Start: start, Length: length, Allocated: true}, img.Readable()
```
Need to fail out of bounds requests like qcow2 version.

nirs replied:
Fixed to always fail since we don't know the image size and the image is not readable.

nirs commented on image/raw/raw.go:
```diff
@@ -14,6 +14,14 @@ type Raw struct {
+	// TODO: For simplificy we report that all extents are allocated and do not
+	// read as zero. This is correct but inefficient. We can implement this using
+	// SEEK_HOLE/SEEK_DATA.
+	return image.Extent{Start: start, Length: length, Allocated: true}, nil
```
- Need to fail out of bounds request like qcow2

nirs replied:
Fixed to validate offset+length like qcow2.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -959,6 +959,122 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+// Next returns the next extent in this image. An extent describes one or more
+// clusters having the same status.
+func (img *Qcow2) Next(start, length int64) (image.Extent, error) {
+	// Default to zero length non-existent cluster, read as zero.
```
Fix comment, this is zero value of image.Extent, but it does not read as zero.

nirs replied:
fixed

nirs commented on image/image.go:
```diff
@@ -8,10 +8,24 @@ import (
 
+type Extent struct {
+	// Offset from start of the image in bytes.
+	Start int64 `json:"start"`
```
This is not strictly needed, since when calling Next(start, length) we always return the requested start value.

nirs replied:
Removing Start makes testing easier and eliminate the error of invalid extents list. When consuming the extents we keep current offset externally so we don't really need it. When printing extents having start is helpful. If we don't report it we have to add another type with Start for encoding json.

nirs commented on image/image.go:
```diff
@@ -8,10 +8,24 @@ import (
+	// Set if this extent is read as zeros.
+	Zero bool `json:"zero"`
+	// Set if this extent is compressed.
+	Compressed bool `json:"compressed"`
```
Allocated and Compressed are not needed for sparse copy, only Zero is needed. They can be useful for other purposes, for example dumping all extents like `qemu-img map --output json`.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -959,6 +959,122 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+
+	for remaining > 0 {
+		clusterNo := offset / int64(img.clusterSize)
+		clusterStart := clusterNo * int64(img.clusterSize)
```
Can more efficient since we always advance in full cluster - no need to compute cluster number for every iteration.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -959,6 +959,122 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+		if current.Length == 0 {
+			// First cluster: copy status to current.
+			current = clusterStatus
+			current.Start = start
```
If start was not aligned to cluster size need to fixup current.Length to remove start-clusterStart.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -959,6 +959,122 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+		// If start+length is not aligned to cluster size, fixup the length.
+		if clusterStatus.Length > remaining {
+			clusterStatus.Length = remaining
+		}
```
This is relevant to the last cluster.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -959,6 +959,122 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+		}
+
+		offset += clusterStatus.Length
+		remaining -= clusterStatus.Length
```
should use current.Length since in the first cluster we may need to advance in smaller value.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -699,56 +699,101 @@ func (img *Qcow2) getL2Table(l1Entry l1TableEntry) ([]l2TableEntry, error) {
+	if cm.L2Entry.compressed() {
+		cm.Compressed = true
+	} else {
+		cm.Zero = standardClusterDescriptor(desc).allZero()
```
Add comment about extended l2 clusters:
cm.Zero is always false. To find which sub cluster is allocated/zero we need to iterate over the allocation bitmap in the extended l2 cluster entry.

nirs replied:
Added comment.

nirs commented on image/qcow2/qcow2.go:
```diff
@@ -962,6 +962,114 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+		Length:     int64(img.clusterSize),
+		Allocated:  true,
+		Compressed: cm.Compressed,
+		Zero:       cm.Zero,
```
Add comment about extended l2 - the cluster is always non-zero. We need to iterate on the sub cluster allocation bitmap to find which subcluster are allocated/zero. Current code is correct but less efficient - we will do very course grain (1 MiB) zero detection for extended l2, instead of handling 32 subclusters in every cluster.

nirs replied:
Commented in getClusterMeta.

AkihiroSuda commented on image/image.go:
```diff
@@ -8,10 +8,24 @@ import (
 // Type must be a "Backing file format name string" that appears in QCOW2.
 type Type string
 
+type Extent struct {
```
Needs godoc to clarify how this is different from clusters

nirs replied:
Right, will add.

AkihiroSuda commented on image/image.go:
```diff
@@ -8,10 +8,24 @@ import (
 type Image interface {
 	io.ReaderAt
 	io.Closer
+	Next(start, length int64) (Extent, error)
```
Does this return EOF on the end of the image?

nirs replied:
No, since you know the size of the image and must use the correct size to use this. If you specify a range which is after the end of the image it will fail.

AkihiroSuda commented on image/stub/stub.go:
```diff
@@ -25,6 +25,11 @@ type Stub struct {
 }
 
+// Next returns the next extent.
+func (img *Stub) Next(start, length int64) (image.Extent, error) {
```
Should this validate `start` and `length`?

nirs replied:
Yes, like the qcow2 version.

nirs replied:
Actually not needed since this always fail now. We cannot report extents when the size of the image is unknown and the image is not readable.

AkihiroSuda commented on convert/convert.go:
```diff
@@ -151,40 +153,57 @@ func (c *Converter) Convert(wa io.WriterAt, ra io.ReaderAt, size int64) error {
-					n := len(buf)
-					if end-start < int64(len(buf)) {
-						n = int(end - start)
+					// Get next extent in this segment.
```
Needs definition of "segment", "area", "extent", "cluster"

nirs replied:
I improved the comments, I hope it is clear enough now.

AkihiroSuda commented on image/image.go:
```diff
@@ -8,10 +8,24 @@ import (
 type Image interface {
 	io.ReaderAt
 	io.Closer
+	Next(start, length int64) (Extent, error)
```
Maybe just
```suggestion
	Extent(start, length int64) (Extent, error)
```

nirs replied:
Can work, but this does give a hint that this is useful for iterating over the image - similar to the tar package.
https://pkg.go.dev/archive/tar#Reader

Another option:

```go
NextExtent(start, length int64) (Extent, error)
```

But it repeats Extent in the name and in the return value.

AkihiroSuda replied:
The difference is that tar.Reader.Next is stateful while Image.Next here seems stateless

nirs replied:
Good point, Next() is stateless. It exposes the structure of the image. It is more like `seek(SEE_HOLE)` or `seek(SEEK_DATA)`.

Lets go with Extent().

nirs replied:
Renamed.

nirs commented:
Current version improves the tests but it not ready yet. I did not make the changes we discussed yet.

I extract the changes preparing for this change to #39 so we can merge them early and make this change smaller.

nirs commented on qcow2reader_test.go:
```diff
@@ -7,15 +7,18 @@ import (
 	"math/rand"
 	"os"
 	"path/filepath"
+	"slices"
```
Not available in go 1.20?

nirs replied:
Replace with reflect.

jandubois replied:
I think it is sensible to bump Go to 1.22

nirs replied:
Update in #39.

nirs commented on qcow2reader_test.go:
```diff
@@ -159,6 +162,361 @@ func BenchmarkExtentsUnallocated(b *testing.B) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !slices.Equal(compressed(extents), actual) {
```
If slices is not available, we can add add an implementation for []image.Extent, or use relect.DeepEqual.

nirs commented on qcow2reader_test.go:
```diff
@@ -159,6 +162,361 @@ func BenchmarkExtentsUnallocated(b *testing.B) {
+	}
+	return nil
+}
+
```
Maybe extract extents related tests to extents_test.go?

nirs replied:
We can improve this later.

nirs commented on convert/convert.go:
```diff
@@ -151,40 +153,57 @@ func (c *Converter) Convert(wa io.WriterAt, ra io.ReaderAt, size int64) error {
+						return
+					}
+					if extent.Zero {
+						start += extent.Length
```
We can update the progress here:

```go
progress.Update(extent.Length)
```

nirs commented on convert/convert.go:
```diff
@@ -151,40 +153,57 @@ func (c *Converter) Convert(wa io.WriterAt, ra io.ReaderAt, size int64) error {
+
+						extent.Length -= int64(nr)
+						extent.Start += int64(nr)
+						start += int64(nr)
```
We can update here for extents read from the image and optionally written to the target here:

```go
progress.Update(nr)
```

This is also more correct than updating the progress in ReadAt, which leads so reporting higher throughput than the actual, since detecting zeros and writing takes some time.

nirs commented:
Last push added test for image with backing file. The new tests can be reviewed but I did not handle yet the rest of the needed changes.

I think the last missing test is an image with shorter backing file.

nirs commented:
Last push add the last missing test for shorter backing file. The new test can be reviewed but I did not handle yet the rest of the needed changes.

nirs commented:
Last push changes:
- Rename Next() to Extent()
- Improve documentation
- Reorder commits to group image and tests changes

Still missing
- documentation
- input validation
- support for progress in convert (ReadeAt and WriteAt cannot be used now)

AkihiroSuda commented on convert/convert.go:
```diff
@@ -78,6 +78,14 @@ func (o *Options) Validate() error {
+	// Called from multiple goroutines after a byte range of length was converted.
+	// If the conversion is successfu, the total number of bytes will be the image
+	// virtual size.
+	Update(n int64)
```
Can n be negative?

nirs replied:
No, but this is the type the go and the kernel use for file size (off_t). It is easier to use when we can use Extent.Length when updating.

We an add a check for n < 0 in the progress bar example implementation.

nirs commented:
Current version addresses all the comments and implements everything.

It is not compatible with lima since convert has additional argument and the reader cannot be used now to report progress. I'll send fix for lima later.

AkihiroSuda reviewed (approved):
Thanks

AkihiroSuda commented on convert/convert.go:
```diff
@@ -129,10 +148,13 @@ func (c *Converter) reset(size int64) {
+// file full of zeroes. To get a sparse target image, the image must be a new
+// empty file, since Convert does not punch holes for zero ranges even if the
+// underlying file system supports hole punching.
+func (c *Converter) Convert(wa io.WriterAt, img image.Image, size int64, progress Updater) error {
```
`progress Updater` could be provided in `Options`?
Then we would not have to change the function signature?

nirs replied:
It is possible, but even if we do we accept now an image instead of a ReaderAt and size.

Also progress scope is one convert, while convert instance can convert many images using the same options, so it is better to accept progress per convert operation.

The options are not something you change per image, it may change depending on the machine (e.g number of cores).

nirs replied:
There other options that we may want to add in the future that may be per image options, so moving the options to Convert makes sense anyway.
